File: kata-trivia-python/trivia/question_deck.py
from collections import deque


class QuestionDeck:
    CATEGORIES = ("Pop", "Science", "Sports", "Rock")

    def __init__(self, questions_per_category):
        self.questions_by_category = {
            category: deque(
                f"{category} Question {index}"
                for index in range(questions_per_category)
            )
            for category in self.CATEGORIES
        }

    def next_question(self, category):
        return self.questions_by_category[category].popleft()

    def category_for(self, position):
        return self.CATEGORIES[(position - 1) % len(self.CATEGORIES)]

File: kata-trivia-java/src/main/java/trivia/QuestionDeck.java
package trivia;

import java.util.HashMap;
import java.util.LinkedList;
import java.util.Map;

class QuestionDeck {
   private final Map<String, LinkedList<String>> questionsByCategory = new HashMap<>();

   QuestionDeck(int questionsPerCategory) {
      addQuestions("Pop", questionsPerCategory);
      addQuestions("Science", questionsPerCategory);
      addQuestions("Sports", questionsPerCategory);
      addQuestions("Rock", questionsPerCategory);
   }

   String nextQuestion(String category) {
      return questionsByCategory.get(category).removeFirst();
   }

   private void addQuestions(String category, int count) {
      LinkedList<String> questions = new LinkedList<>();
      for (int index = 0; index < count; index++) {
         questions.addLast(category + " Question " + index);
      }
      questionsByCategory.put(category, questions);
   }
}
